File: packages/sdk-python/lantern/runtime/llm_client.py
"""Production LLM client that talks to the model router over gRPC.

In the Lantern runtime, LLM calls are routed through the model router
which handles caching, routing to concrete models, metering, and
multi-tenant isolation.
"""

from __future__ import annotations

import json
import os
from typing import Any, AsyncIterator

from lantern.errors import LanternLlmError
from lantern.types import Capability, LlmClient as LlmClientBase, Message


class GrpcLlmClient(LlmClientBase):
    """LLM client that communicates with the model router via gRPC.

    Initialized with a gRPC channel from the runtime.
    """

    def __init__(self, grpc_endpoint: str, run_id: str, tenant_id: str) -> None:
        self._endpoint = grpc_endpoint
        self._run_id = run_id
        self._tenant_id = tenant_id
        self._channel: Any = None

    async def _ensure_channel(self) -> Any:
        if self._channel is None:
            try:
                import grpc.aio  # type: ignore[import-untyped]
                self._channel = grpc.aio.insecure_channel(self._endpoint)
            except ImportError:
                raise LanternLlmError(
                    "grpcio is required for production runtime LLM calls. "
                    "Install with: pip install grpcio"
                )
        return self._channel

    async def complete(self, **kwargs: Any) -> str:
        """Send a completion request to the model router.

        Keyword Args:
            prompt: Simple string prompt (convenience).
            messages: List of Message dicts for multi-turn.
            capability: Model capability to request.
            optimize: Optimization target.
            tools: Tool definitions for function calling.
            max_tokens: Maximum output tokens.
            temperature: Sampling temperature.
            stop: Stop sequences.
            no_cache: Bypass response cache.
        """
        channel = await self._ensure_channel()

        # Build the gRPC request payload
        request = self._build_request(kwargs)

        try:
            # In production this calls the generated proto stub.
            # For now we use a generic unary call pattern that will be
            # wired to the generated lantern.modelrouter.v1 service.
            from google.protobuf import json_format, struct_pb2  # type: ignore[import-untyped]

            request_struct = struct_pb2.Struct()
            json_format.ParseDict(request, request_struct)

            # The actual gRPC call — stub will be generated from protos
            # lantern.modelrouter.v1.ModelRouter/Complete
            stub = channel.unary_unary(
                "/lantern.modelrouter.v1.ModelRouter/Complete",
                request_serializer=request_struct.SerializeToString,
                response_deserializer=struct_pb2.Struct.FromString,
            )

            metadata = [
                ("x-lantern-tenant-id", self._tenant_id),
                ("x-lantern-run-id", self._run_id),
            ]

            response = await stub(request_struct, metadata=metadata)
            result = json_format.MessageToDict(response)
            return result.get("content", "")

        except Exception as exc:
            if "grpc" in type(exc).__module__:
                raise LanternLlmError(
                    f"Model router gRPC error: {exc}",
                    capability=kwargs.get("capability"),
                ) from exc
            raise

    async def json(self, **kwargs: Any) -> Any:
        """Send a completion request that returns structured JSON."""
        kwargs.setdefault("response_format", "json")
        raw = await self.complete(**kwargs)
        try:
            return json.loads(raw)
        except json.JSONDecodeError as exc:
            raise LanternLlmError(f"Failed to parse LLM JSON response: {exc}") from exc

    async def stream(self, **kwargs: Any) -> AsyncIterator[str]:
        """Stream tokens from the model router."""
        channel = await self._ensure_channel()
        request = self._build_request(kwargs)

        try:
            from google.protobuf import json_format, struct_pb2  # type: ignore[import-untyped]

            request_struct = struct_pb2.Struct()
            json_format.ParseDict(request, request_struct)

            stub = channel.unary_stream(
                "/lantern.modelrouter.v1.ModelRouter/StreamComplete",
                request_serializer=request_struct.SerializeToString,
                response_deserializer=struct_pb2.Struct.FromString,
            )

            metadata = [
                ("x-lantern-tenant-id", self._tenant_id),
                ("x-lantern-run-id", self._run_id),
            ]

            async for chunk in stub(request_struct, metadata=metadata):
                result = json_format.MessageToDict(chunk)
                delta = result.get("delta", "")
                if delta:
                    yield delta

        except Exception as exc:
            if "grpc" in type(exc).__module__:
                raise LanternLlmError(
                    f"Model router stream gRPC error: {exc}",
                    capability=kwargs.get("capability"),
                ) from exc
            raise

    async def embed(
        self, texts: list[str], capability: Capability | None = None
    ) -> list[list[float]]:
        """Generate embeddings via the model router."""
        channel = await self._ensure_channel()

        try:
            from google.protobuf import json_format, struct_pb2  # type: ignore[import-untyped]

            request = {
                "texts": texts,
                "capability": capability or "embed-large",
            }
            request_struct = struct_pb2.Struct()
            json_format.ParseDict(request, request_struct)

            stub = channel.unary_unary(
                "/lantern.modelrouter.v1.ModelRouter/Embed",
                request_serializer=request_struct.SerializeToString,
                response_deserializer=struct_pb2.Struct.FromString,
            )

            metadata = [
                ("x-lantern-tenant-id", self._tenant_id),
                ("x-lantern-run-id", self._run_id),
            ]

            response = await stub(request_struct, metadata=metadata)
            result = json_format.MessageToDict(response)
            return result.get("embeddings", [])

        except Exception as exc:
            if "grpc" in type(exc).__module__:
                raise LanternLlmError(f"Embed gRPC error: {exc}") from exc
            raise

    async def close(self) -> None:
        if self._channel is not None:
            await self._channel.close()
            self._channel = None

    @staticmethod
    def _build_request(kwargs: dict[str, Any]) -> dict[str, Any]:
        """Convert keyword args to a gRPC request dict."""
        request: dict[str, Any] = {}

        if "prompt" in kwargs:
            request["messages"] = [{"role": "user", "content": kwargs["prompt"]}]
        elif "messages" in kwargs:
            messages = kwargs["messages"]
            request["messages"] = [
                m.model_dump() if hasattr(m, "model_dump") else m
                for m in messages
            ]

        for key in ("capability", "optimize", "max_tokens", "temperature", "stop", "no_cache", "response_format"):
            if key in kwargs and kwargs[key] is not None:
                request[key] = kwargs[key]

        if "tools" in kwargs and kwargs["tools"]:
            request["tools"] = [
                t.model_dump() if hasattr(t, "model_dump") else t
                for t in kwargs["tools"]
            ]

        return request

File: packages/sdk-ts/src/runtime/llm-client.ts
/**
 * Production LLM client that routes all model calls through the
 * runtime sidecar to the Lantern model router.
 *
 * The model router handles:
 * - Capability-to-model resolution (e.g. "reasoning-large" -> claude-opus)
 * - Vendor failover and load balancing
 * - Semantic caching
 * - Token metering and cost tracking
 * - Rate limiting per tenant
 *
 * The SDK never talks to a model vendor directly.
 */

import type {
  LlmClient,
  LlmOptions,
  LlmJsonOptions,
  LlmStreamOptions,
  Capability,
  Message,
  ToolDef,
  GroundedResult,
} from "../types.js";
import type { Runtime } from "./runtime.js";
import type {
  CompleteRequest,
  GrpcMessage,
  GrpcToolDef,
  RequestMeta,
} from "./grpc-client.js";
import { LanternLlmError, LanternLlmJsonError } from "./errors.js";
import { traced } from "./tracing.js";

// ---------------------------------------------------------------------------
// Message / tool conversion
// ---------------------------------------------------------------------------

/** Convert SDK Message to gRPC message format. */
function toGrpcMessage(msg: Message): GrpcMessage {
  return {
    role: msg.role,
    content: msg.content,
    name: msg.name,
    toolCalls: msg.toolCalls?.map((tc) => ({
      id: tc.id,
      name: tc.name,
      arguments: tc.arguments,
    })),
    toolCallId: msg.toolCallId,
  };
}

/** Convert SDK ToolDef to gRPC tool definition format. */
function toGrpcToolDef(td: ToolDef): GrpcToolDef {
  return {
    name: td.name,
    description: td.description,
    parametersJson: JSON.stringify(td.parameters),
  };
}

/** Build the messages array for a CompleteRequest. */
/**
 * Build the ground-or-abstain system contract for a grounded completion.
 * Pure + exported so it's unit-testable without a runtime sidecar.
 */
export function buildGroundingContract(sources: string[]): string {
  const list = sources.length
    ? sources.map((s, i) => `[${i + 1}] ${s}`).join("\n")
    : "(none provided)";
  return (
    "GROUND-OR-ABSTAIN: Assert ONLY what the SOURCES below support. " +
    "If the answer is not in the sources, say you don't know or state it as intent — never as fact. " +
    "Never invent names, numbers, dates, statuses, or completed actions. Cite the source you used.\n\n" +
    "SOURCES:\n" +
    list
  );
}

function buildMessages(opts: LlmOptions): GrpcMessage[] {
  if (opts.messages) {
    return opts.messages.map(toGrpcMessage);
  }
  if (opts.prompt) {
    return [{ role: "user", content: opts.prompt }];
  }
  throw new LanternLlmError("Either 'prompt' or 'messages' must be provided");
}

/** Resolve the capability string, defaulting to "auto". */
function resolveCapability(opts: LlmOptions): string {
  return opts.capability ?? "auto";
}

/** Resolve the optimize target to a string for the gRPC request. */
function resolveOptimize(opts: LlmOptions): string | undefined {
  if (!opts.optimize) return undefined;
  if (typeof opts.optimize === "string") return opts.optimize;
  return JSON.stringify(opts.optimize);
}

// ---------------------------------------------------------------------------
// RuntimeLlmClient
// ---------------------------------------------------------------------------

/** Maximum number of retries for JSON schema validation failures. */
const JSON_VALIDATION_MAX_RETRIES = 3;

/**
 * Production LLM client that sends all requests through the runtime
 * sidecar's model router interface.
 */
export class RuntimeLlmClient implements LlmClient {
  private readonly runtime: Runtime;

  constructor(runtime: Runtime) {
    this.runtime = runtime;
  }

  /**
   * Send a completion request and return the full text response.
   *
   * @param opts - Completion options (prompt or messages, capability, etc.).
   * @returns The model's text response.
   * @throws {LanternLlmError} If the model call fails.
   */
  async complete(opts: LlmOptions): Promise<string> {
    const sidecar = this.runtime.sidecar;
    if (!sidecar) {
      throw new LanternLlmError("LLM client requires a production runtime with sidecar");
    }

    const meta: RequestMeta = { ...this.runtime.meta };
    const capability = resolveCapability(opts);

    return traced("llm.complete", { capability }, async (span) => {
      const req: CompleteRequest = {
        meta,
        messages: buildMessages(opts),
        capability,
        optimize: resolveOptimize(opts),
        tools: opts.tools?.map(toGrpcToolDef),
        maxTokens: opts.maxTokens,
        temperature: opts.temperature,
        stop: opts.stop,
        noCache: opts.noCache,
      };

      try {
        const res = await sidecar.complete(req);
        span.setAttribute("model", res.model);
        span.setAttribute("tokensIn", res.tokensIn);
        span.setAttribute("tokensOut", res.tokensOut);
        span.setAttribute("costUsd", res.costUsd);
        return res.text;
      } catch (err) {
        if (err instanceof LanternLlmError) throw err;
        throw new LanternLlmError(
          err instanceof Error ? err.message : String(err),
          undefined,
          true,
        );
      }
    });
  }

  /**
   * Complete under a ground-or-abstain contract. The model is instructed to
   * assert only what the supplied `sources` support and to say it doesn't know
   * (or state intent) rather than fabricate. Returns the reply plus the sources
   * it was grounded against. The platform's server-side action guard still runs
   * on top; this adds the citation contract at the agent's own boundary.
   */
  async completeGrounded(
    opts: LlmOptions & { sources: string[] },
  ): Promise<GroundedResult> {
    const { sources = [], prompt, messages, ...rest } = opts;
    const contract = buildGroundingContract(sources);
    const base: Message[] = messages
      ? [...messages]
      : prompt
        ? [{ role: "user", content: prompt }]
        : [];
    const text = await this.complete({
      ...rest,
      messages: [{ role: "system", content: contract }, ...base],
    });
    return { text, sources, grounded: sources.length > 0 };
  }

  /**
   * Send a completion request with `response_format: json`, validate the
   * response against a Zod schema, and return the parsed object.
   *
   * If the model output fails validation, the request is retried up to
   * {@link JSON_VALIDATION_MAX_RETRIES} times with the validation errors
   * appended as a correction prompt.
   *
   * @param opts - Completion options including a Zod schema.
   * @returns The parsed and validated response object.
   * @throws {LanternLlmJsonError} If validation fails after all retries.
   * @throws {LanternLlmError} If the model call itself fails.
   */
  async json<T>(opts: LlmJsonOptions<T>): Promise<T> {
    const sidecar = this.runtime.sidecar;
    if (!sidecar) {
      throw new LanternLlmError("LLM client requires a production runtime with sidecar");
    }

    const meta: RequestMeta = { ...this.runtime.meta };
    const capability = resolveCapability(opts);

    return traced("llm.json", { capability }, async (span) => {
      let messages = buildMessages(opts);
      let lastRaw = "";
      let lastValidationErrors = "";

      for (let attempt = 0; attempt < JSON_VALIDATION_MAX_RETRIES; attempt++) {
        const req: CompleteRequest = {
          meta,
          messages,
          capability,
          optimize: resolveOptimize(opts),
          tools: opts.tools?.map(toGrpcToolDef),
          maxTokens: opts.maxTokens,
          temperature: opts.temperature,
          stop: opts.stop,
          responseFormat: "json",
          noCache: opts.noCache || attempt > 0, // Don't cache retries
        };

        let rawText: string;
        try {
          const res = await sidecar.complete(req);
          rawText = res.text;
          span.setAttribute("model", res.model);
          span.setAttribute("tokensIn", res.tokensIn);
          span.setAttribute("tokensOut", res.tokensOut);
        } catch (err) {
          if (err instanceof LanternLlmError) throw err;
          throw new LanternLlmError(
            err instanceof Error ? err.message : String(err),
            undefined,
            true,
          );
        }

        lastRaw = rawText;

        // Parse JSON
        let parsed: unknown;
        try {
          parsed = JSON.parse(rawText);
        } catch {
          lastValidationErrors = `Response is not valid JSON: ${rawText.slice(0, 500)}`;
          // Append correction and retry
          messages = [
            ...messages,
            { role: "assistant", content: rawText },
            {
              role: "user",
              content: `Your response was not valid JSON. Please try again and respond with only valid JSON. Error: ${lastValidationErrors}`,
            },
          ];
          continue;
        }

        // Validate against Zod schema
        const result = opts.schema.safeParse(parsed);
        if (result.success) {
          span.setAttribute("attempts", attempt + 1);
          return result.data;
        }

        // Validation failed: build error message and retry
        lastValidationErrors = result.error.issues
          .map((issue) => `${issue.path.join(".")}: ${issue.message}`)
          .join("; ");

        messages = [
          ...messages,
          { role: "assistant", content: rawText },
          {
            role: "user",
            content: `Your JSON response did not match the expected schema. Please fix the following issues and try again:\n${lastValidationErrors}`,
          },
        ];
      }

      // All retries exhausted
      throw new LanternLlmJsonError(lastRaw, lastValidationErrors);
    });
  }

  /**
   * Send a streaming completion request and yield text deltas as an
   * async iterator.
   *
   * The stream is end-to-end: tokens flow from the model through the
   * sidecar to the SDK without buffering. Each yielded string is a
   * text delta (not the full response so far).
   *
   * @param opts - Completion options.
   * @yields Text delta strings.
   * @throws {LanternLlmError} If the stream fails.
   */
  async *stream(opts: LlmStreamOptions): AsyncIterable<string> {
    const sidecar = this.runtime.sidecar;
    if (!sidecar) {
      throw new LanternLlmError("LLM client requires a production runtime with sidecar");
    }

    const meta: RequestMeta = { ...this.runtime.meta };
    const capability = resolveCapability(opts);

    const req: CompleteRequest = {
      meta,
      messages: buildMessages(opts),
      capability,
      optimize: resolveOptimize(opts),
      tools: opts.tools?.map(toGrpcToolDef),
      maxTokens: opts.maxTokens,
      temperature: opts.temperature,
      stop: opts.stop,
      noCache: opts.noCache,
    };

    try {
      for await (const chunk of sidecar.streamComplete(req)) {
        if (chunk.delta) {
          yield chunk.delta;
        }
        if (chunk.done) {
          return;
        }
      }
    } catch (err) {
      if (err instanceof LanternLlmError) throw err;
      throw new LanternLlmError(
        err instanceof Error ? err.message : String(err),
        undefined,
        true,
      );
    }
  }

  /**
   * Compute embeddings for one or more texts.
   *
   * @param texts - The texts to embed.
   * @param capability - Embedding capability (defaults to "embed-large").
   * @returns A 2D array of vectors, one per input text.
   * @throws {LanternLlmError} If the embedding call fails.
   */
  async embed(texts: string[], capability?: Capability): Promise<number[][]> {
    const sidecar = this.runtime.sidecar;
    if (!sidecar) {
      throw new LanternLlmError("LLM client requires a production runtime with sidecar");
    }

    const meta: RequestMeta = { ...this.runtime.meta };
    const cap = capability ?? "embed-large";

    return traced("llm.embed", { capability: cap, count: texts.length }, async (span) => {
      try {
        const res = await sidecar.embed({ meta, texts, capability: cap });
        span.setAttribute("model", res.model);
        span.setAttribute("tokensIn", res.tokensIn);
        return res.vectors;
      } catch (err) {
        if (err instanceof LanternLlmError) throw err;
        throw new LanternLlmError(
          err instanceof Error ? err.message : String(err),
          undefined,
          true,
        );
      }
    });
  }
}
